amp-form XHR test is flaky

```
...................................
Chrome 58.0.3029 (Linux 0.0.0)  shadow ampdoc  amp-form   XHR   AMP-Redirect-To should redirect on error and header is set FAILED
	AssertionError: expected 0 to equal 1
	    at /tmp/0d63eb9846d3beddb9e442e31c71c7cb.browserify:73241:38 <- /home/travis/build/ampproject/amphtml/extensions/amp-form/0.1/test/test-amp-form.js:1560:37
	    at <anonymous>
```

https://travis-ci.org/ampproject/amphtml/builds/232622496?utm_source=github_status&utm_medium=notification

--- a/extensions/amp-form/0.1/amp-form.js
+++ b/extensions/amp-form/0.1/amp-form.js
@@ -29,7 +29,7 @@ import {
   isProxyOrigin,
   parseUrl,
 } from '../../../src/url';
-import {dev, user, rethrowAsync} from '../../../src/log';
+import {dev, user} from '../../../src/log';
 import {getMode} from '../../../src/mode';
 import {xhrFor} from '../../../src/services';
 import {toArray} from '../../../src/types';
@@ -475,27 +475,24 @@ export class AmpForm {
       this.renderTemplate_(json || {});
       this.maybeHandleRedirect_(response);
     }, error => {
-      const message = 'Failed to parse response JSON:';
-      user().error(TAG, message, error);
-      rethrowAsync(message, error);
+      user().error(TAG, `Failed to parse response JSON: ${error}`);
     });
   }
 
   /**
    * Transition the form the the submit error state.
-   * @param {../../../src/service/xhr-impl.FetchError} error
+   * @param {../../../src/service/xhr-impl.FetchError} errorResponse
    * @private
    */
-  handleXhrSubmitFailure_(error) {
+  handleXhrSubmitFailure_(errorResponse) {
+    const error = (errorResponse && errorResponse.error) || errorResponse;
     this.triggerAction_(
-        /* success */ false, error ? error.responseJson : null);
+        /* success */ false, errorResponse ? errorResponse.responseJson : null);
     this.analyticsEvent_('amp-form-submit-error');
     this.setState_(FormState_.SUBMIT_ERROR);
-    this.renderTemplate_(error.responseJson || {});
-    this.maybeHandleRedirect_(error.response);
-    const message = 'Form submission failed:';
-    user().error(TAG, message, error);
-    rethrowAsync(message, error);
+    this.renderTemplate_(errorResponse.responseJson || {});
+    this.maybeHandleRedirect_(errorResponse.response);
+    user().error(TAG, `Form submission failed: ${error}`);
   }
 
   /** @private */

--- a/extensions/amp-form/0.1/test/test-amp-form.js
+++ b/extensions/amp-form/0.1/test/test-amp-form.js
@@ -40,6 +40,7 @@ import {
 import {FetchError} from '../../../../src/service/xhr-impl';
 import '../../../amp-selector/0.1/amp-selector';
 import {toggleExperiment} from '../../../../src/experiments';
+import {user} from '../../../../src/log';
 
 describes.repeated('', {
   'single ampdoc': {ampdoc: 'single'},
@@ -758,7 +759,8 @@ describes.repeated('', {
         expect(form.className).to.not.contain('amp-form-submit-error');
         expect(form.className).to.not.contain('amp-form-submit-success');
         fetchResolver({json: () => Promise.resolve()});
-        return timer.promise(5).then(() => {
+
+        return ampForm.xhrSubmitPromiseForTesting().then(() => {
           expect(ampForm.actions_.trigger).to.be.called;
           expect(ampForm.actions_.trigger).to.be.calledWith(form, 'submit');
           expect(ampForm.state_).to.equal('submit-success');
@@ -771,6 +773,8 @@ describes.repeated('', {
               /** CustomEvent */ sinon.match.has('detail'));
           expect(ampForm.analyticsEvent_).to.be.calledWith(
               'amp-form-submit-success');
+        }, () => {
+          expect(false).to.be.true; // should not run
         });
       });
     });
@@ -804,7 +808,10 @@ describes.repeated('', {
         expect(form.className).to.not.contain('amp-form-submit-error');
         expect(form.className).to.not.contain('amp-form-submit-success');
         fetchRejecter();
-        return timer.promise(5).then(() => {
+
+        return ampForm.xhrSubmitPromiseForTesting().then(() => {
+          expect(false).to.be.true; // should not run
+        }, () => {
           expect(button1.hasAttribute('disabled')).to.be.false;
           expect(button2.hasAttribute('disabled')).to.be.false;
           expect(ampForm.state_).to.equal('submit-error');
@@ -1505,7 +1512,8 @@ describes.repeated('', {
           sandbox.stub(ampForm.xhr_, 'fetch').returns(fetchResolvePromise);
           redirectToValue = 'https://google.com/';
           ampForm.handleSubmitAction_();
-          return timer.promise(10).then(() => {
+
+          return ampForm.xhrSubmitPromiseForTesting().then(() => {
             expect(env.win.top.location.href).to.be.equal(redirectToValue);
           });
         });
@@ -1514,7 +1522,10 @@ describes.repeated('', {
           sandbox.stub(ampForm.xhr_, 'fetch').returns(fetchResolvePromise);
           redirectToValue = 'http://google.com/';
           ampForm.handleSubmitAction_();
-          return timer.promise(10).then(() => {
+
+          return ampForm.xhrSubmitPromiseForTesting().then(() => {
+            expect(false).to.be.true; // should not run
+          }, () => {
             expect(env.win.top.location.href).to.be.equal(
                 'https://example-top.com/');
           });
@@ -1524,7 +1535,10 @@ describes.repeated('', {
           sandbox.stub(ampForm.xhr_, 'fetch').returns(fetchResolvePromise);
           redirectToValue = '/hello';
           ampForm.handleSubmitAction_();
-          return timer.promise(10).then(() => {
+
+          return ampForm.xhrSubmitPromiseForTesting().then(() => {
+            expect(false).to.be.true; // should not run
+          }, () => {
             expect(env.win.top.location.href).to.be.equal(
                 'https://example-top.com/');
           });
@@ -1535,7 +1549,10 @@ describes.repeated('', {
           sandbox.stub(ampForm.xhr_, 'fetch').returns(fetchResolvePromise);
           redirectToValue = 'http://google.com/';
           ampForm.handleSubmitAction_();
-          return timer.promise(10).then(() => {
+
+          return ampForm.xhrSubmitPromiseForTesting().then(() => {
+            expect(false).to.be.true; // should not run
+          }, () => {
             expect(env.win.top.location.href).to.be.equal(
                 'https://example-top.com/');
           });
@@ -1544,21 +1561,13 @@ describes.repeated('', {
         it('should redirect on error and header is set', () => {
           sandbox.stub(ampForm.xhr_, 'fetch').returns(fetchRejectPromise);
           redirectToValue = 'https://example2.com/hello';
-          const errors = [];
-          const realSetTimeout = window.setTimeout;
-          sandbox.stub(window, 'setTimeout', (callback, delay) => {
-            realSetTimeout(() => {
-              try {
-                callback();
-              } catch (e) {
-                errors.push(e);
-              }
-            }, delay);
-          });
+          const logSpy = sandbox.spy(user(), 'error');
           ampForm.handleSubmitAction_();
-          return timer.promise(10).then(() => {
-            expect(errors.length).to.equal(1);
-            expect(errors[0]).to.match(/Form submission failed/);
+
+          return ampForm.xhrSubmitPromiseForTesting().then(() => {
+            expect(logSpy).to.be.calledOnce;
+            const error = logSpy.getCall(0).args[1];
+            expect(error).to.match(/Form submission failed/);
             expect(env.win.top.location.href).to.be.equal(redirectToValue);
           });
         });

--- a/extensions/amp-form/0.1/test/test-amp-form.js
+++ b/extensions/amp-form/0.1/test/test-amp-form.js
@@ -774,7 +774,7 @@ describes.repeated('', {
           expect(ampForm.analyticsEvent_).to.be.calledWith(
               'amp-form-submit-success');
         }, () => {
-          expect(false).to.be.true; // should not run
+          assert.fail('Submit should have succeeded.');
         });
       });
     });
@@ -810,7 +810,7 @@ describes.repeated('', {
         fetchRejecter();
 
         return ampForm.xhrSubmitPromiseForTesting().then(() => {
-          expect(false).to.be.true; // should not run
+          assert.fail('Submit should have failed.');
         }, () => {
           expect(button1.hasAttribute('disabled')).to.be.false;
           expect(button2.hasAttribute('disabled')).to.be.false;
@@ -1524,7 +1524,7 @@ describes.repeated('', {
           ampForm.handleSubmitAction_();
 
           return ampForm.xhrSubmitPromiseForTesting().then(() => {
-            expect(false).to.be.true; // should not run
+            assert.fail('Submit should have failed.');
           }, () => {
             expect(env.win.top.location.href).to.be.equal(
                 'https://example-top.com/');
@@ -1537,7 +1537,7 @@ describes.repeated('', {
           ampForm.handleSubmitAction_();
 
           return ampForm.xhrSubmitPromiseForTesting().then(() => {
-            expect(false).to.be.true; // should not run
+            assert.fail('Submit should have failed.');
           }, () => {
             expect(env.win.top.location.href).to.be.equal(
                 'https://example-top.com/');
@@ -1551,7 +1551,7 @@ describes.repeated('', {
           ampForm.handleSubmitAction_();
 
           return ampForm.xhrSubmitPromiseForTesting().then(() => {
-            expect(false).to.be.true; // should not run
+            assert.fail('Submit should have failed.');
           }, () => {
             expect(env.win.top.location.href).to.be.equal(
                 'https://example-top.com/');

